rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -1127,17 +1127,24 @@ extern "C" rmw_ret_t rmw_init(const rmw_init_options_t * options, rmw_context_t
   context->implementation_identifier = eclipse_cyclonedds_identifier;
   context->impl = nullptr;
 
-  if ((ret = rmw_init_options_copy(options, &context->options)) != RMW_RET_OK) {
-    return ret;
+  ret = rmw_init_options_copy(options, &context->options);
+  if (RMW_RET_OK != ret) {
+    goto fail;
   }
 
-  rmw_context_impl_t * impl = new(std::nothrow) rmw_context_impl_t();
-  if (nullptr == impl) {
-    return RMW_RET_BAD_ALLOC;
+  context->impl = new(std::nothrow) rmw_context_impl_t();
+  if (nullptr == context->impl) {
+    ret = RMW_RET_BAD_ALLOC;
+    goto fail;
   }
-
-  context->impl = impl;
-  return RMW_RET_OK;
+  return ret;
+fail:
+  if (RMW_RET_OK != rmw_init_options_fini(&context->options)) {
```

Comment:
Is it safe to call `fini` if `init` failed? Just checking ...

Reply:
good point, I think it's not.
@hidmic maybe just returning an error in line 1132.

Reply:
Well, arguably it should not and either `rmw_init_options_init()` or `rmw_init_options_copy()` should guarantee output options remain uninitialized. Reality is that such guarantee is not documented in `rmw` API and most implementations do not really check. `rmw_cyclonedds_cpp` isn't the exception. Thus the extra `fini`, which judging by the code isn't unsafe (though again, not necessarily correct nor ideal).

Alternatively, we can establish that guarantee and change implementations to abide to it. Thoughts?

Reply:
> Alternatively, we can establish that guarantee and change implementations to abide to it. Thoughts?

:+1:
If I call `*_init()` or `*_copy()` and the function fails, I expect it to cleanup all the resources it created (and not having to do it manually).
In the case of `*_copy(src, dst)`, I also expect the function to no modify `*dst` if it failed to do the copy.

It's the more usual pattern in C code IMO.

Reply:
See https://github.com/ros2/rmw/pull/244.